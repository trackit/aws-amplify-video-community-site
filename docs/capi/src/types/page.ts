import {HyperscriptNode} from "./hyperscript-node";
import {Menu} from "./menu";

export type Filters = Record<string, string[]>;

export interface PageLink {
  route: string;
  title: string;
  filters?: Filters;
}

export type Direction = "next" | "previous";

export interface Page extends PageLink {
  sectionTitle?: string;
  productRootLink?: PageLink;
  relativeToContentDir: string;
  versions?: Record<string, string>;
  route: string;
  description: string;
  body: HyperscriptNode[];
  menu?: Menu;
  filters?: Filters;
  disableTOC?: boolean;
  noTemplate?: boolean;
  filterKey?: string;
  next?: Record<string, PageLink> | PageLink;
  previous?: Record<string, PageLink> | PageLink;
}
