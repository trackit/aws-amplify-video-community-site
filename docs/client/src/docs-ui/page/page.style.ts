import {css} from "emotion";
import {MAX_WIDTH} from "../../amplify-ui/styles/media";

export const sidebarLayoutDivStyle = css`
  display: flex;
  flex-direction: row;
  max-width: ${MAX_WIDTH};
  margin: 0 auto;
  background-color: var(--bg-color);
`;

export const sidebarLayoutStyle = css`
  display: flex;
  flex-direction: column;
  background-color: var(--bg-color-tertiary);

  h1 {
    margin-top: 0.375rem;
  }

  p {
    margin-bottom: 1rem;

    &.searchable-code {
      display: none;
    }
  }

  a {
    h2,
    h3 {
      color: var(--font-color);
    }
  }

  amplify-toc-contents h4 {
    margin-bottom: 1rem;
  }

  amplify-sidebar-layout-toggle {
    position: fixed;
    right: 1rem;
    bottom: 1rem;
    background-color: var(--color-orange-hv);
    border-radius: 50%;
    width: 3.5rem;
    height: 3.5rem;

    img {
      width: 1.5rem;
      height: 1.5rem;
    }
  }
`;

export const sidebarHeaderStyle = css`
  display: flex;
  flex-direction: row;
  margin: 1.75rem 2.75rem 0 2.5rem;
`;

export const mainStyle = css`
  a {
    &,
    &:h2,
    &:h3 {
      color: var(--font-color);
    }

    &:hover,
    & h2:hover,
    & h3:hover {
      text-decoration: underline;
    }
  }
`;

export const pageStyle = css`
  position: relative;
  display: flex;
  flex-direction: column;

  ul {
    margin-bottom: 1rem;

    li + li {
      margin-top: 0.25rem;
    }
  }
`;

export const sectionHeaderStyle = css`
  line-height: normal;
  margin-bottom: 0;
  font-size: 1rem;
  text-transform: uppercase;
  color: var(--font-color-secondary);
  font-weight: bold;
`;

export const tocStyle = css`
  > div {
    position: relative;
    height: 100%;

    > amplify-toc {
      position: sticky;
      top: 3rem;
      flex-direction: column;
      max-height: 100vh;
      overflow-y: auto;

      > div {
        padding-bottom: 3rem;
      }
    }
  }
`;
