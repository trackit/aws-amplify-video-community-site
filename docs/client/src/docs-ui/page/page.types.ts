import {
  PLATFORM_FILTER_OPTIONS,
  integrationFilterOptions,
  FRAMEWORK_FILTER_OPTIONS,
} from "../../utils/filter-data";

export interface SelectedFilters {
  platform?: typeof PLATFORM_FILTER_OPTIONS[number];
  framework?: typeof FRAMEWORK_FILTER_OPTIONS[number];
  integration?: typeof integrationFilterOptions[number];
}

export type SetSelectedFilters = (updates: SelectedFilters) => void;

export type SelectedTabHeadings = string[];
export type SetNewSelectedTabHeadings = (tabHeading: string) => void;

export interface PageContext {
  alwaysRerenderBlockSwitcher?: number;
  selectedFilters?: SelectedFilters;
  setSelectedFilters?: SetSelectedFilters;
  selectedTabHeadings: SelectedTabHeadings;
  setNewSelectedTabHeadings: SetNewSelectedTabHeadings;
}
