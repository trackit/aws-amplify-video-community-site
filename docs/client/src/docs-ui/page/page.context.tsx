import {createProviderConsumer} from "@stencil/state-tunnel";
import {h} from "@stencil/core";
import {PageContext} from "./page.types";

export const pageContext = createProviderConsumer<PageContext>(
  {} as PageContext,
  (subscribe, child) => <context-consumer {...{subscribe}} renderer={child} />,
);
