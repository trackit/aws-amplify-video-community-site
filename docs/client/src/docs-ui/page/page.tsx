import {Component, Host, h, State, Listen, Element, Build} from "@stencil/core";
import {
  sidebarLayoutStyle,
  sidebarLayoutDivStyle,
  pageStyle,
  tocStyle,
  mainStyle,
  sectionHeaderStyle,
  sidebarHeaderStyle,
} from "./page.style";
import {
  Page,
  createVNodesFromHyperscriptNodes,
  filtersByRoute,
} from "../../api";
import {updateDocumentHead} from "../../utils/update-document-head";
import {
  getFilterKeyFromPage,
  getFilterKeyFromLocalStorage,
  withFilterOverrides,
} from "../../utils/filters";
import {filterOptionsByName} from "../../utils/filter-data";
import {
  SetSelectedFilters,
  SelectedTabHeadings,
  SetNewSelectedTabHeadings,
} from "./page.types";
import {pageContext} from "./page.context";
import {
  track,
  trackPageVisit,
  trackPageFetchException,
  AnalyticsEventType,
} from "../../utils/track";
import {ensureMenuScrolledIntoView} from "../../utils/ensure-menu-scrolled-into-view";
import {getPage} from "../../cache.worker";
import {getNavHeight} from "../../utils/get-nav-height";
import {scrollToHash} from "../../utils/scroll-to-hash";
import {parseURL} from "../../utils/url/url.worker";
import {redirects} from "./redirects";

const SELECTED_TABS_LOCAL_STORAGE_KEY = `amplify-docs::selected-tabs`;

@Component({tag: "docs-page", shadow: false})
export class DocsPage {
  @Element() el: HTMLElement;

  @State() pageData?: Page;
  @State() blendUniversalNav?: boolean;
  @State() sidebarStickyTop = getNavHeight("rem");
  @State() selectedFilters: Record<string, string | undefined> = {};
  @State() selectedTabHeadings: SelectedTabHeadings = [];

  rafId?: number;
  isFirstRenderOfCurrentPage = true;
  previousPathname = "";
  /** increments every time the page changes, to tell all block switchers on the page to rerender */
  alwaysRerenderBlockSwitcher = 0;

  setNewSelectedTabHeading: SetNewSelectedTabHeadings = (tabHeading) => {
    // create temp array with `tabHeading` (the new highest priority) as first el
    const nextSelectedTabHeadings = new Array<string>();
    nextSelectedTabHeadings.push(tabHeading);

    // iterate through previous `selectedTabHeadings`
    this.selectedTabHeadings.forEach((e) => {
      // no repeats allowed!
      if (tabHeading !== e) {
        // ensure preexisting tab name priorities are preserved
        nextSelectedTabHeadings.push(e);
      }
    });

    // set the new priority list in state
    this.selectedTabHeadings = nextSelectedTabHeadings;

    // and serialize and save it to local storage
    localStorage.setItem(
      SELECTED_TABS_LOCAL_STORAGE_KEY,
      JSON.stringify(this.selectedTabHeadings),
    );
  };

  setSelectedFilters: SetSelectedFilters = (updates) => {
    const overrides = withFilterOverrides(updates, this.selectedFilters);
    this.selectedFilters = {
      ...this.selectedFilters,
      ...overrides,
    };
    for (const [filterKey, filterValue] of Object.entries(overrides)) {
      localStorage.setItem(
        getFilterKeyFromLocalStorage(filterKey),
        filterValue,
      );
    }
  };

  filterKey?: string;
  filterValue?: string;
  validFilterValue = true;

  @Listen("resize", {target: "window"})
  setSidebarStickyTop() {
    if (Build.isBrowser && this.pageData?.menu) {
      this.sidebarStickyTop = getNavHeight("rem");
    }
  }

  startRaf() {
    // lets us repeatedly call `updatePageData` without actually
    // triggering redundant `this.getPageData` calls.
    let {pathname} = location;
    const updatePageData = () => {
      if (location.pathname !== pathname) {
        this.isFirstRenderOfCurrentPage = true;
        this.previousPathname = pathname;
        pathname = location.pathname;
        this.getPageData();
      }
    };

    // create RAF loop, save its ID (so that we can end the loop in `componentWillUnload`).
    // This loop triggers `updatePageData`, which––upon path changes––triggers the appropriate rerender.
    this.rafId = (function watchForRouteChange() {
      updatePageData();
      return requestAnimationFrame(watchForRouteChange);
    })();
  }

  componentDidLoad() {
    this.startRaf();
  }

  scrollToId() {
    const {hash} = location;
    if (hash) {
      setTimeout(() => {
        // TODO: fix potential race condition
        scrollToHash(hash, this.el);
      }, 250);
    }
  }

  componentDidRender() {
    this.setSidebarStickyTop();
    if (this.isFirstRenderOfCurrentPage) {
      this.scrollToId();
    }
    this.isFirstRenderOfCurrentPage = false;
  }

  stopRaf() {
    if (this.rafId) {
      cancelAnimationFrame(this.rafId);
    }
  }

  componentWillUnload() {
    this.stopRaf();
  }

  restoreBlockSwitcherState() {
    // gather list of previously-selected tab headings (might be null)
    const persistedSelectedTabsSerialized =
      localStorage.getItem(SELECTED_TABS_LOCAL_STORAGE_KEY) || undefined;
    if (persistedSelectedTabsSerialized) {
      // save that selection array if it exists (otherwise, list is empty)
      this.selectedTabHeadings = JSON.parse(persistedSelectedTabsSerialized);
    }
  }

  async componentWillLoad() {
    this.restoreBlockSwitcherState();
    return await this.getPageData();
  }

  async getPageData() {
    let currentRoute = location.pathname || "/";
    if (!currentRoute.startsWith("/")) {
      currentRoute = `/${currentRoute}`;
    }
    if (currentRoute.endsWith("/") && currentRoute !== "/") {
      currentRoute = currentRoute.substring(0, currentRoute.length - 1);
    }
    if (currentRoute.includes("cli/graphql-transformer/directives")) {
      const redirected = redirects[location.hash];
      if (redirected) {
        currentRoute = `/cli/graphql-transformer${redirected}`;
        // @ts-ignore
        window.location = currentRoute;
      }
    }

    const {path, params} = await parseURL(currentRoute);
    const routeFiltersEntry = filtersByRoute.get(path);
    const allFilters =
      routeFiltersEntry &&
      Object.values(routeFiltersEntry).reduce((acc, curr) => {
        return [...acc, ...curr];
      }, []);
    this.blendUniversalNav = currentRoute === "/";

    if (Build.isBrowser) {
      track({
        type: AnalyticsEventType.PAGE_VISIT,
        attributes: {
          url: currentRoute,
          previousUrl: this.previousPathname,
          referrer: document.referrer,
        },
      });
      trackPageVisit();
    }

    try {
      const pageData = await getPage(currentRoute);
      if (!pageData) {
        trackPageFetchException();
        this.pageData = undefined;
      } else {
        this.pageData = pageData;
        updateDocumentHead(pageData);
        this.filterKey = getFilterKeyFromPage(pageData);
        this.selectedFilters = Object.assign(
          {},
          ...Object.keys(filterOptionsByName).map((filterKey) => {
            const localStorageKey = getFilterKeyFromLocalStorage(filterKey);
            return {
              [filterKey]: localStorageKey
                ? localStorage.getItem(localStorageKey) || undefined
                : undefined,
            };
          }),
        );

        if (this.filterKey) {
          const {[this.filterKey]: filterValue} = params;
          if (typeof filterValue === "string" && filterValue !== "undefined") {
            this.filterValue = filterValue;
            if (allFilters) {
              this.validFilterValue = allFilters.includes(filterValue);
              if (this.validFilterValue) {
                this.setSelectedFilters({[this.filterKey]: this.filterValue});
              }
            }
          } else {
            this.filterValue = undefined;
          }
        } else {
          this.filterKey = undefined;
        }
      }
    } catch (exception) {
      track({
        type: AnalyticsEventType.PAGE_DATA_FETCH_EXCEPTION,
        attributes: {url: location.href, exception},
      });
      trackPageFetchException();
    }
  }

  showMenu = (): boolean => {
    const menuItems = this.pageData?.menu;
    return !!(menuItems && menuItems.length > 0);
  };

  render() {
    return (
      <Host class={pageStyle}>
        <pageContext.Provider
          state={{
            alwaysRerenderBlockSwitcher: this.alwaysRerenderBlockSwitcher++,
            selectedFilters: this.selectedFilters,
            setSelectedFilters: this.setSelectedFilters,
            selectedTabHeadings: this.selectedTabHeadings,
            setNewSelectedTabHeadings: this.setNewSelectedTabHeading,
          }}
        >
          {this.blendUniversalNav ? (
            <docs-universal-nav-blend
              heading="Amplify Docs"
              brand-icon="/assets/logo-dark.svg"
            ></docs-universal-nav-blend>
          ) : (
            <docs-universal-nav
              heading="Amplify Docs"
              brand-icon="/assets/logo-light.svg"
            ></docs-universal-nav>
          )}
          {this.pageData && this.pageData.noTemplate
            ? createVNodesFromHyperscriptNodes(this.pageData.body)
            : [
                <docs-secondary-nav
                  pageHasMenu={!!this.pageData && !!this.pageData.menu}
                />,
                this.pageData && this.validFilterValue ? (
                  <div class={sidebarLayoutStyle}>
                    <amplify-toc-provider>
                      <amplify-sidebar-layout>
                        <div class={sidebarLayoutDivStyle}>
                          {this.showMenu() && (
                            <amplify-sidebar-layout-sidebar
                              top={this.sidebarStickyTop}
                            >
                              <div>
                                <div class={sidebarHeaderStyle}>
                                  <amplify-sidebar-close-button />
                                  {this.pageData?.filterKey && (
                                    <docs-select-anchor page={this.pageData} />
                                  )}
                                </div>
                                <docs-menu
                                  filterKey={this.filterKey}
                                  page={this.pageData}
                                  key={this.pageData?.productRootLink?.route}
                                />
                              </div>
                            </amplify-sidebar-layout-sidebar>
                          )}
                          <amplify-sidebar-layout-main class={mainStyle}>
                            <amplify-toc-contents>
                              <div>
                                {this.pageData && [
                                  <h1
                                    class={{
                                      [sectionHeaderStyle]: true,
                                      "category-heading": true,
                                    }}
                                  >
                                    {this.pageData.sectionTitle}
                                  </h1>,
                                  <h1 class="page-heading">
                                    {this.pageData.title}
                                  </h1>,
                                  createVNodesFromHyperscriptNodes(
                                    this.pageData.body,
                                  ),
                                  <docs-next-previous
                                    key={this.pageData.route}
                                    page={this.pageData}
                                  />,
                                ]}
                              </div>
                            </amplify-toc-contents>
                            <amplify-sidebar-layout-toggle
                              onClick={ensureMenuScrolledIntoView}
                              in-view-class="in-view"
                              class="three-dee-effect"
                            >
                              <img
                                class="burger-graphic"
                                src="/assets/burger.svg"
                              />
                              <img class="ex-graphic" src="/assets/close.svg" />
                            </amplify-sidebar-layout-toggle>
                          </amplify-sidebar-layout-main>
                          {!this.pageData?.disableTOC && (
                            <div class={tocStyle}>
                              <div>
                                <amplify-toc pageTitle={this.pageData?.title} />
                              </div>
                            </div>
                          )}
                        </div>
                      </amplify-sidebar-layout>
                    </amplify-toc-provider>
                  </div>
                ) : (
                  <docs-four-o-four />
                ),
                <docs-footer />,
              ]}
        </pageContext.Provider>
      </Host>
    );
  }
}
