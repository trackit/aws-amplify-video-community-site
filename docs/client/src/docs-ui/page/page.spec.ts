import {DocsPage} from "./page";
import {newSpecPage} from "@stencil/core/testing";

describe("docs-page", () => {
  describe("Component logic", () => {
    let docsPage: DocsPage;
    beforeEach(() => (docsPage = new DocsPage()));

    it("should init `selected` as `{}`", () =>
      expect(docsPage.selectedFilters).toEqual({}));

    it("should init `data` as `undefined`", () =>
      expect(docsPage.pageData).toBeUndefined());

    it("should init `blendUniversalNav` as `undefined`", () =>
      expect(docsPage.blendUniversalNav).toBeUndefined());
  });

  describe("Render logic", () => {
    it("should render", async () => {
      expect(
        (
          await newSpecPage({
            components: [DocsPage],
            html: `<docs-page />`,
          })
        ).root,
      ).toMatchSnapshot();
    });
  });
});
