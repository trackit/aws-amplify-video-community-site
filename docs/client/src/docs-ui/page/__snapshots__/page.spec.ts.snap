// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docs-page Render logic should render 1`] = `
<docs-page class="css-ntiofr">
  <docs-universal-nav-blend brand-icon="/assets/logo-dark.svg" heading="Amplify Docs"></docs-universal-nav-blend>
  <docs-secondary-nav></docs-secondary-nav>
  <div class="css-hj6f3l">
    <amplify-toc-provider>
      <amplify-sidebar-layout>
        <div class="css-1oqiadb">
          <amplify-sidebar-layout-main class="css-1j9n6qe">
            <amplify-toc-contents>
              <div>
                <h1 class="category-heading css-6tygts"></h1>
                <h1 class="page-heading"></h1>
                <docs-next-previous></docs-next-previous>
              </div>
            </amplify-toc-contents>
            <amplify-sidebar-layout-toggle class="three-dee-effect" in-view-class="in-view">
              <img class="burger-graphic" src="/assets/burger.svg">
              <img class="ex-graphic" src="/assets/close.svg">
            </amplify-sidebar-layout-toggle>
          </amplify-sidebar-layout-main>
          <div class="css-34jkc1">
            <div>
              <amplify-toc></amplify-toc>
            </div>
          </div>
        </div>
      </amplify-sidebar-layout>
    </amplify-toc-provider>
  </div>
  <docs-footer></docs-footer>
</docs-page>
`;
